### ISSUE TITLE
include filter returns an undefined object instead of empty array

### ISSUE THREAD

**Issue by trosen333** [fuwaneko/loopback-connector-rethinkdb#10]
When using include filter for defined relation if none of the objects in database have ids for their relations defined an undefined object will be returned - which crashes server. it should instead return an empty array. 

Steps for reproduction:
1. Clone https://github.com/svarogg/loopback-example-relations
2. Configure some rethinkdb datasource in server/datasources.json. Make sure it contains the tables required for loopback to work. (in this case r.expr(["User", "AccessToken", "ACL", "RoleMapping", "Role", "Customer", "Review", "Order"]).forEach(r.tableCreate(r.row)))
3. npm install && slc run
4. Go to http://localhost:3000/explorer
5. Set some access token
6. Send request Reviews/Test
7. See in console output: TypeError: Cannot read property 'length' of undefined

Control:
1. Change server/boot/sample.js. uncomment line 51 (ie add a Review object to the db that contains an orderID) :
         Q.all([
            ///uncomment the first review in order to resolve issue
            Q.ninvoke(Review, "create", {product: 'Product1', star: 3, order: orderA}),
            Q.ninvoke(Review, "create", {product: 'Product2', star: 2}),
            Q.ninvoke(Review, "create", {product: 'Product5', star: 5})
          ])
2. Repeat steps 3-6 from reproduction
3. Notice that console output has no errors and request returns all Review objects.

Thanks in advance for any assistance.

**Comment by fuwaneko** [comment]
Could you post stacktrace?

**Comment by trosen333** [comment]
events.js:72
        throw er; // Unhandled 'error' event
              ^
TypeError: Cannot read property 'length' of undefined
    at Object.async.each (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:152:17)
    at linkOneToMany (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/lib/include.js:757:19)
    at async.each.callback (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:157:13)
    at _each (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:57:9)
    at Object.async.each (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:156:9)
    at targetLinkingTask (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/lib/include.js:753:17)
    at results (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:723:21)
    at _asyncMap (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:391:17)
    at async.each.callback (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:157:13)
    at _each (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:57:9)
    at async.each (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:156:9)
    at _asyncMap (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:390:13)
    at Object.doParallel [as map] (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:361:23)
    at _parallel (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:721:20)
    at Object.async.parallel (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:751:9)
    at module.exports (/Users/tali/loopback-example-relations/node_modules/loopback-datasource-juggler/lib/include.js:99:11)

Process finished with exit code 7

**Comment by trosen333** [comment]
Just wanted to update you on this issue:
1. I've updated my example code, should be easier to run now if you couldn't get it before.
2. We were able to narrow down the issue a little bit more:
It's a problem with BelongsTo relationship. It seems as though there is some issue with the inq: 

- in order to see the issue in the code go to: loopback-example-relations ▸ node_modules ▸ loopback-datasource-juggler ▸ lib▸ include.js  (line: 753  async.each(targets, linkOneToMany, next);)

- the targetFetchHandler function tries to run code on an object that is undefined (line: 753)  
objTargetIdMap ends up being dropped

- we think it might have something to do with  filter.where[relation.keyTo] = {
        inq: uniq(targetIds)
      };(line 728)

not much substance, i know, but it's definitely clear the relationship is broken.

**Comment by fuwaneko** [comment]
@trosen333 Still no idea what's going on. The example is kinda complicated as it has subincludes, i.e. include orders and customers for orders. Async and lack of proper docs on "how to write datasource connector" does not help in debugging either.

**Comment by AvivLander** [comment]
@bajtos, could we summon you to possibly help with this issue? We desperately want to use ReThinkDB in a LoopBackJS project, but without this issue resolved, it's simply not usable. There's a real problem understanding what is going on here. I'm certain anyone using ReThinkDB will face this issue of BelongsTo relationship setups.

**Comment by fuwaneko** [comment]
@AvivLander it's not only belongsTo, but all relationships in general.

**Comment by fuwaneko** [comment]
@AvivLander I switched async from 1.0.0 in loopback-datasource-juggler to 1.3.0 and — boom — the problem is solved. I remember that there were some problems with async in loopback (like loosing context, etc) so I decided to give it a try.

Open loopback-datasource-juggler's package.json in node_modules of your example project and manually change async version from 1.0.0 to 1.3.0, run npm install and then try to run project.

**Comment by AvivLander** [comment]
@fuwaneko you are the man!

I have a call tonight with some of the people at StrongLoop. I'll be sure to mention this to them.

**Comment by fuwaneko** [comment]
@AvivLander that doesn't solve all problems though, relations are still kinda broken, but I've no idea what to do. I believe it's related to ID fields somehow, would be great if you mention this too. In RethinkDB ids are UUIDs treated as strings in driver and I'm not sure how to make datasource-juggler to know that.

**Comment by altsang** [comment]
@superkhau - can you pls take a look?

**Comment by superkhau** [comment]
@altsang Sure, can you prioritize it on waffle pls?

**Comment by altsang** [comment]
Will do

**Comment by altsang** [comment]
@AvivLander i've created a story for @superkhau to address, but FYI that it's doubtful that he'll be able to look at it in the next coming sprint because of customer issues he's working on that we actually put off till now, just wanted to let you know in case your timeline is critical

**Comment by bajtos** [comment]
@AvivLander sorry for not responding to your comment, I was on a paternity leave.

**Comment by perfectedTech** [comment]
@bajtos - congratulations!

**Comment by slyadams** [comment]
Has there been any movement on this front? I am having to make all my FE services work around the fact that the if there are no existing relations, the included relation key doesn't exist, rather than being an empty array.

**Comment by fuwaneko** [comment]
@slyadams if you can provide pull requests, I'll gladly merge them.

**Comment by slyadams** [comment]
Sure, I can take a look. Never looked at the gubbins before, could you give me a pointer in which module handles this piece?

**Comment by fuwaneko** [comment]
@slyadams relations are handled inside loopback-datasource-juggler.

**Comment by Westermann** [comment]
Any updates on this? I am experiencing the same problem.

**Comment by bajtos** [comment]
Hello, sorry for the delay. I am not familiar with the code in `lib/include.js` myself. However, when I added a bit of logging around the failing code, and compared the output between a run using RethinkDB connector and a run using built-in Memory connector, I observed that the following query is returning different results: `Order.find({"where":{"id":{"inq":[]}}})`

 - With memory connector, the query returns an empty array (as expected, because `inq` is empty)
 - With RethinkDB connector, a list of orders is returned. It looks like all orders are returned, but I didn't verify.

I managed to track the issue down to [`operators.inq`](https://github.com/fuwaneko/loopback-connector-rethinkdb/blob/46636d8494bbc0c3526368665dc1b26f933c123c/rethink.js#L675-L687), where `inq('id', [])` returns `undefined` instead of `[]`.

I believe this is a bug that needs to get fixed in the RethinkDB connector. Any volunteers to contribute the patch?

As it turns out, juggler did not have any tests covering the `inq` operator, I opened a pull request to add them - see https://github.com/strongloop/loopback-datasource-juggler/pull/1232. The second new test should start failing in the current implementation of the RethinkDB connector, providing a failing test for the TDD reg-green-refactor cycle.
